src/main/java/com/suse/saltstack/netapi/client/SaltStackJdkConnection.java

```diff
@@ -0,0 +1,110 @@
+ */
+public class SaltStackJdkConnection implements SaltStackConnection {
+
+    /** The endpoint. */
+    private String endpoint;
+
+    /** The config object. */
+    private final SaltStackClientConfig config;
+
+    /**
+     * Init a connection to a given SaltStack API endpoint.
+     *
+     * @param endpointIn the endpoint
+     * @param configIn the config
+     */
+    public SaltStackJdkConnection(String endpointIn, SaltStackClientConfig configIn) {
+        endpoint = endpointIn;
+        config = configIn;
+    }
+
+    /**
+     * Send a POST request and parse the result into object of given {@link Type}.
+     *
+     * @param resultType the type of the result
+     * @param data the data to send (in JSON format)
+     * @return object of type given by resultType
+     * @throws SaltStackException if the request was not successful
+     */
+    public <T> T getResult(Type resultType, String data) throws SaltStackException {
+        return request(resultType, "POST", data);
+    }
+
+    /**
+     * Perform HTTP request and parse the result into a given result type.
+     *
+     * @param resultType the type of the result
+     * @param method the HTTP method to use
+     * @return object of type given by resultType
+     * @throws SaltStackException in case of a problem
+     */
+    private <T> T request(Type resultType, String method, String data)
+            throws SaltStackException {
+        HttpURLConnection connection = null;
+        InputStream inputStream = null;
+        try {
+            // Setup and configure the connection
+            connection = SaltStackRequestFactory.getInstance().initConnection(
+                    method, endpoint, config);
+            connection.setUseCaches(false);
+            connection.setDoInput(true);
+            connection.setDoOutput(true);
+
+            // Send data in case we have some
+            if (data != null) {
+                connection.setRequestProperty("Content-Type", "application/json");
+                connection.setRequestProperty("Content-Length",
+                        Integer.toString(data.getBytes().length));
+
+                // Actually write the bytes
+                DataOutputStream outputStream = new DataOutputStream(
+                        connection.getOutputStream());
+                outputStream.writeBytes(data);
+                outputStream.flush();
+                outputStream.close();
+            }
+
+            // React depending on the result of the request
+            int responseCode = connection.getResponseCode();
+            if (responseCode == HttpURLConnection.HTTP_OK ||
+                    responseCode == HttpURLConnection.HTTP_ACCEPTED) {
+                inputStream = connection.getInputStream();
+                Reader inputStreamReader = new InputStreamReader(inputStream);
+                Reader streamReader = new BufferedReader(inputStreamReader);
```

Comment:
Input streams has to be always closed, especially in `HttpURLConnection` because there are lots of things happening internally in there, such as keep-alive, connection buffering options etc.